XiEditor/Fps.swift

```diff
@@ -0,0 +1,179 @@
+        //
+        // frames per second = 1 / (seconds per frame)
+        //                   === 1 / (nanoseconds per frame / nanoseconds per second)
+        //                   === nanoseconds per second / nanoseconds per frame
+        //                   === (ns/s) / (ns/frame)
+        //                   === (ns/s) * (frame/ns)
+        return Double(NSEC_PER_SEC) / Double(nanosecondsPerFrame)
+    }
+
+    public func fps() -> Double {
+        return FpsMeasurement.fps(nanosecondsPerFrame: self.elapsedNanoseconds())
+    }
+
+    public func elapsedSeconds() -> Double {
+        return Double(self.elapsedNanoseconds()) / Double(NSEC_PER_SEC)
+    }
+
+    public func elapsedMilliseconds() -> Double {
+        return Double(self.elapsedNanoseconds()) / Double(NSEC_PER_MSEC)
+    }
+
+    public func elapsedNanoseconds() -> UInt64 {
+        return self.end.uptimeNanoseconds - self.start.uptimeNanoseconds
+    }
+
+    public func elapsedNanoseconds(since start: DispatchTime) -> UInt64 {
+        return end.uptimeNanoseconds - start.uptimeNanoseconds
+    }
+
+    public func elapsedSeconds(since start: DispatchTime) -> Double {
+        return Double(self.elapsedNanoseconds(since: start)) / Double(NSEC_PER_SEC)
+    }
+}
+
+public class FpsTimer {
+    private let fps : Fps
+    private let start : DispatchTime
+
+    fileprivate init(fps: Fps) {
+        self.fps = fps
+        self.start = DispatchTime.now()
+    }
+
+    deinit {
+        let end = DispatchTime.now()
+        self.fps.save(measurement: FpsMeasurement(from: start, to: end))
+    }
+}
+
+public struct FpsSnapshot {
+    private let samples: [FpsMeasurement]
+
+    fileprivate init(samples: [FpsMeasurement]) {
+        // Sort by increasing FPS (samples[0] is slowest frame, samples[len] is
+        // fastest frame).
+        self.samples = samples.sorted(by: { (fps1, fps2) -> Bool in
+            return fps1.elapsedNanoseconds() >= fps2.elapsedNanoseconds()
+        })
+    }
+
+    public func minFps() -> Double {
+        return samples.first?.fps() ?? Double.nan
+    }
+
+    public func maxFps() -> Double {
+        return samples.last?.fps() ?? Double.nan
+    }
+
+    public func fps(percentile: Double) -> Double {
+        if samples.count == 0 {
+            return Double.nan
+        }
+
+        var sample_idx = Int(Double(samples.count) * percentile)
+        if sample_idx >= samples.count {
+           sample_idx = samples.count - 1
+        }
+        return samples[sample_idx].fps()
+    }
+
+    public func meanFps() -> Double {
+        let totalNanoseconds = samples.reduce(UInt64(0)) { (elapsedNanosTotal, measurement) -> UInt64 in
+            return elapsedNanosTotal + measurement.elapsedNanoseconds()
+        }
+        let meanNanosecondsPerFrame = totalNanoseconds / UInt64(samples.count)
+        return FpsMeasurement.fps(nanosecondsPerFrame: meanNanosecondsPerFrame)
+    }
+}
+
+public protocol FpsObserver : class {
```

Comment:
Do we need both of these? I saw only one was actually implemented. Or maybe the fps could be an optional field that we only include when there's a tick?

Reply:
I wanted to do it as an overlay view but I'm a bit lazy to hook that part up.    You did get that the changed with the raw frame rate happens every tick but the FpsSnapshot is a longer time period (i.e. the past 1 second).